In [ ]:
import numpy as np
import pandas as pd
import requests
from google.colab import drive
drive.mount('/content/drive')
import time

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


A. Write a Python program which takes as its argument5 the path to a text file of data (assumed to be in the itemset format above) and produces as output to the console a sparse ARFF file.

In [ ]:
start_time = time.time()

In [ ]:
url = "http://fimi.uantwerpen.be/data/kosarak.dat"
response = requests.get(url)
data = response.text
lines = data.split('\n')

In [ ]:
int_lines = []
for line in lines:
  curr_line = []
  for item in line.split():
    curr_line.append(int(item))
  int_lines.append(curr_line)

In [ ]:
int_lines.pop()

[]

In [ ]:
max_id = 0
for line in int_lines:
  max_id = max(max_id, max(line))

In [ ]:
print(f"Number of instances: {len(int_lines)} and Number of attributes: {max_id}")

Number of instances: 990002 and Number of attributes: 41270


EXAMPLE<br>
1 2 3<br>
1<br>
4 5 6 7<br>
1 8<br>
9 10<br>

After Conversion: <br>
@relation sparseData<br>
@attribute 1 numeric<br>
@attribute 2 numeric<br>
@attribute 3 numeric<br>
@attribute 4 numeric<br>
@attribute 5 numeric<br>
@attribute 6 numeric<br>
@attribute 7 numeric<br>
@attribute 8 numeric<br>
@attribute 9 numeric<br>
@attribute 10 numeric<br>
@data<br>
{0 1,1 1,2 1}<br>
{0 1}<br>
{3 1,4 1,5 1,6 1}<br>
{0 1,7 1}<br>
{8 1,9 1}
<br>
CSV Counterpart (as generated by weka): <br>
1,2,3,4,5,6,7,8,9,10<br>
1,1,1,0,0,0,0,0,0,0<br>
1,0,0,0,0,0,0,0,0,0<br>
0,0,0,1,1,1,1,0,0,0<br>
1,0,0,0,0,0,0,1,0,0<br>
0,0,0,0,0,0,0,0,1,1<br>


In [ ]:
filename = "/content/drive/MyDrive/USML/HW1/PROB_2/cleanedSparseClickData.arff"

In [ ]:
def write_arff(filename, data, max_id):
  header = f"@relation cleanedSparseClickData\n"

  for i in range(1, max_id+1):
    header += f"@attribute {i} "
    header += "{0, 1}\n"

  header += "@data\n"

  data_section = ""
  for line in data:
    line = list(set(line)) # done for the purpose of data cleaning
    line.sort()
    data_section += "{"
    for i in range(len(line)-1):
      data_section += f"{line[i]-1} 1,"
    data_section += f"{line[-1]-1} 1"
    data_section += "}\n"

  data_section = data_section[:-1] # Remove the last newline character


  with open(filename, "w", encoding="utf-8") as file:
    file.write(header)
    file.write(data_section)

In [ ]:
write_arff(filename, int_lines, max_id)

In [ ]:
end_time = time.time()
print(f"Time taken: {end_time - start_time}")

Time taken: 42.63153266906738


B. Use your program to convert the kosarak.dat file to a sparse kosarak.arff. About how long did it take to run?<br>
Ans: About 43 sec <br>

<br>
Additional Note on Data Cleaning:<br>
{0 1,5 1,6 1,10 1,26 1,63 1,82 1,84 1,85 1,85 1,86 1,87 1,210 1,275 1,277 1,320 1,896 1,941 1,1139 1,1528 1,1867 1,1977 1,2156 1,2230 1,2241 1,2312 1,2862 1,2867 1,2881 1,3858 1,6749 1,6812 1,10128 1}
<br>
In this line (51374th instance) index 85 appears twice. Multiple such lines exist in the dataset. <br><br>

C. Load the resulting file into Weka (as described above; you should have 41,270 attributes and 990, 002 instances). About how long did it take to load this file? <br>
Ans: 12 sec
<br><br>
D. Use Weka’s FP-Growth implementation to find rules that have support count of at least 49, 500 and confidence of at least 99% – record your rules (there should be 2).<br>
Ans: <br>
=== Run information ===

Scheme:       weka.associations.FPGrowth -P 1 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0 -S
Relation:     cleanedSparseClickData
Instances:    990002
Attributes:   41270
              [list of attributes omitted]
=== Associator model (full training set) ===

FPGrowth found 2 rules

1. [11=0, 218=0, 148=0]: 50098 ==> [6=0]: 49866   <conf:(1)> lift:(1.64) lev:(0.02) conv:(84.4)
2. [11=0, 148=0]: 55759 ==> [6=0]: 55230   <conf:(0.99)> lift:(1.63) lev:(0.02) conv:(41.3)

<br><br>
E. Run the algorithm at least 5 times. Then look to the log and record how much time each took. How does the average time compare to the time necessary to convert the dataset and then load into Weka?<br>
Ans:

Log: <br>
17:47:26: Weka Explorer<br>
17:47:26: (c) 1999-2022 The University of Waikato, Hamilton, New Zealand<br>
17:47:26: web: https://www.cs.waikato.ac.nz/~ml/weka/<br>
17:47:26: Started on Tuesday, 14 May 2024<br>
<br>
17:47:30: Base relation is now cleanedSparseClickData (990002 instances)<br>
17:48:25: Started weka.associations.FPGrowth<br>
17:48:25: Command: weka.associations.FPGrowth -P 1 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0 -S<br>
17:48:26: Finished weka.associations.FPGrowth<br>
<br>
17:49:59: Base relation is now cleanedSparseClickData (990002 instances)<br>
17:50:38: Started weka.associations.FPGrowth<br>
17:50:38: Command: weka.associations.FPGrowth -P 2 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0<br>
17:50:41: Finished weka.associations.FPGrowth<br>
<br>
17:51:43: Base relation is now cleanedSparseClickData (990002 instances)<br>
17:52:17: Started weka.associations.FPGrowth<br>
17:52:17: Command: weka.associations.FPGrowth -P 1 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0<br>
17:52:20: Finished weka.associations.FPGrowth<br>
<br>
17:52:35: Base relation is now cleanedSparseClickData (990002 instances)<br>
17:53:07: Started weka.associations.FPGrowth<br>
17:53:09: Command: weka.associations.FPGrowth -P 1 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0 -S<br>
17:53:09: Finished weka.associations.FPGrowth<br>
<br>
17:53:25: Base relation is now cleanedSparseClickData (990002 instances)<br>
17:53:58: Started weka.associations.FPGrowth<br>
17:53:58: Command: weka.associations.FPGrowth -P 1 -I -1 -N 10 -T 0 -C 0.99 -D 0.05 -U 1.0 -M 49500.0 -S<br>
17:53:59: Finished weka.associations.FPGrowth<br>

Avg time: 1+3+3+2+1 / 5 = 2 sec


